File: server/data/schemas.go
package data

import (
	"github.com/curio-research/keystone/state"
)

// MAKE SURE TO ADD 'id' FIELD TO ALL SCHEMAS!!!

// randomness variable
type LocalRandSeedSchema struct {
	Id        int `gorm:"primaryKey;autoIncrement:false"`
	RandValue int
}

type Weather int

const (
	Sunny Weather = 1
	Windy Weather = 2
)

type GameSchema struct {
	Id      int `gorm:"primaryKey;autoIncrement:false"`
	GameId  string
	Weather Weather
}

type Terrain bool

const (
	Ground   Terrain = true
	Obstacle Terrain = false
)

type TileSchema struct {
	Id       int       `gorm:"primaryKey;autoIncrement:false"`
	Position state.Pos `gorm:"embedded"`
	Terrain  Terrain
}

type PlayerSchema struct {
	Id              int       `gorm:"primaryKey;autoIncrement:false"`
	Position        state.Pos `gorm:"embedded"`
	Resources       int
	PlayerId        int
	Base64PublicKey string
}

type ProjectileSchema struct {
	Id       int       `gorm:"primaryKey;autoIncrement:false"`
	Position state.Pos `gorm:"embedded"`
}

type AnimalSchema struct {
	Id       int       `gorm:"primaryKey;autoIncrement:false"`
	Position state.Pos `gorm:"embedded"`
}

type ResourceSchema struct {
	Id       int       `gorm:"primaryKey;autoIncrement:false"`
	Position state.Pos `gorm:"embedded"`
	Amount   int
}

// ----------------------------
//
//	table accessors
//
// ----------------------------

var (
	Game            = state.NewTableAccessor[GameSchema]()
	LocalRandomSeed = state.NewTableAccessor[LocalRandSeedSchema]()
	Projectile      = state.NewTableAccessor[ProjectileSchema]()
	Tile            = state.NewTableAccessor[TileSchema]()
	Player          = state.NewTableAccessor[PlayerSchema]()
	Animal          = state.NewTableAccessor[AnimalSchema]()
	Resource        = state.NewTableAccessor[ResourceSchema]()
)

var SchemaMapping = map[interface{}]*state.TableBaseAccessor[any]{
	&GameSchema{}:          (*state.TableBaseAccessor[any])(Game),
	&LocalRandSeedSchema{}: (*state.TableBaseAccessor[any])(LocalRandomSeed),
	&ProjectileSchema{}:    (*state.TableBaseAccessor[any])(Projectile),
	&TileSchema{}:          (*state.TableBaseAccessor[any])(Tile),
	&PlayerSchema{}:        (*state.TableBaseAccessor[any])(Player),
	&AnimalSchema{}:        (*state.TableBaseAccessor[any])(Animal),
	&ResourceSchema{}:      (*state.TableBaseAccessor[any])(Resource),
}

File: client/src/core/schemas.ts
import 'reflect-metadata';
import { TableAccessor } from '../keystone/tableAccessor';

// ------------------------
// Game world schemas
// ------------------------

interface LocalRandSeedSchema {
  RandValue: number;
  Id: number;
}

export enum Weather {
  Sunny = 1,
  Windy = 2,
}

interface GameSchema {
  Id: number;
  Weather: Weather;
}

interface TileSchema {
  Id: number;
  Position: Position;
  Terrain: boolean; // true: ground, false: obstacle
}

interface PlayerSchema {
  Id: number;
  Position: Position;
  Resources: number;
  PlayerId: number;
}

interface ProjectileSchema {
  Id: number;
  Position: Position;
}

interface AnimalSchema {
  Id: number;
  Position: Position;
}

interface ResourceSchema {
  Id: number;
  Position: Position;
  Amount: number;
}

export interface Position {
  x: number;
  y: number;
}

// ---------------------------
// table accessors
// ---------------------------

export const PlayerTable = new TableAccessor<PlayerSchema>('PlayerSchema');
export const TileTable = new TableAccessor<TileSchema>('TileSchema');
export const GameTable = new TableAccessor<GameSchema>('GameSchema');
export const AnimalTable = new TableAccessor<AnimalSchema>('AnimalSchema');
export const ProjectileTable = new TableAccessor<ProjectileSchema>('ProjectileSchema');
export const LocalRandSeedTable = new TableAccessor<LocalRandSeedSchema>('LocalRandSeedSchema');
export const ResourceTable = new TableAccessor<ResourceSchema>('ResourceSchema');

export const AllTableAccessors: TableAccessor<any>[] = [
  PlayerTable,
  TileTable,
  GameTable,
  AnimalTable,
  ProjectileTable,
  LocalRandSeedTable,
  ResourceTable,
];
